geoana/em/static/halfspace.py

```diff
@@ -0,0 +1,733 @@
+
+    def current_density(self, xyz_m, xyz_n=None):
+        """Current density for a dipole source in a halfspace.
+
+       This method computes the current density for a dipole source in a halfspace at
+        the set of gridded xyz locations provided. Where :math:`\\rho`
+        is the electric resistivity and :math:`\\mathbf{E}` is the electric field
+        for the dipole source.
+        The current density :math:`\\mathbf{J}` is:
+
+       .. math::
+
+            \\mathbf{J} = \\frac{\\mathbf{E}}{\\rho}
+
+        Parameters
+        ----------
+        xyz_m : (..., 3) numpy.ndarray
+            Location of the M voltage electrode.
+        xyz_n : (..., 3) numpy.ndarray, optional
+            Location of the N voltage electrode.
+
+        Returns
+        -------
+        J : (..., 3) np.ndarray
+            Current density of point current in units :math:`\\frac{A}{m^2}`.
+
+        Examples
+        --------
+        Here, we define a dipole source in a halfspace to compute current density.
+
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>> from mpl_toolkits.axes_grid1 import make_axes_locatable
+        >>> from geoana.em.static import DipoleHalfSpace
+
+        Define the dipole source.
+
+        >>> rho = 1.0
+        >>> current = 1.0
+        >>> location_a = np.r_[-1, 0, 0]
+        >>> location_b = np.r_[1, 0, 0]
+        >>> simulation = DipoleHalfSpace(
+        >>>     current=current, rho=rho, location_a=location_a, location_b=location_b
+        >>> )
+
+        Now we create a set of gridded locations and compute the current density.
+
+        >>> X1, Y1 = np.meshgrid(np.linspace(-1, 1, 20), np.linspace(-1, 1, 20))
+        >>> X2, Y2 = np.meshgrid(np.linspace(-2, 2, 20), np.linspace(-2, 2, 20))
+        >>> Z = np.zeros_like(X1)
+        >>> xyz1 = np.stack((X1, Y1, Z), axis=-1)
+        >>> xyz2 = np.stack((X2, Y2, Z), axis=-1)
+        >>> j1 = simulation.electric_field(xyz_m=xyz1, xyz_n=None)
+        >>> j2 = simulation.electric_field(xyz_m=xyz1, xyz_n=xyz2)
+
+        Finally, we plot the current density.
+
+        >>> fig, axs = plt.subplots(1, 2, figsize=(18,12))
+        >>> titles = ['3 Electrodes', '4 Electrodes']
+        >>> for ax, J, title in zip(axs.flatten(), [j1, j2], titles):
+        >>>     J_amp = np.linalg.norm(J, axis=-1)
+        >>>     im = ax.pcolor(X1, Y1, J_amp, shading='auto')
+        >>>     divider = make_axes_locatable(ax)
+        >>>     cax = divider.append_axes("right", size="5%", pad=0.05)
+        >>>     cb = plt.colorbar(im, cax=cax)
+        >>>     cb.set_label(label= 'Current Density ($A/m^2$)')
+        >>>     ax.streamplot(X1, Y1, J[..., 0], J[..., 1], density=0.75)
+        >>>     ax.set_ylabel('Y coordinate ($m$)')
+        >>>     ax.set_xlabel('X coordinate ($m$)')
+        >>>     ax.set_aspect('equal')
+        >>>     ax.set_title(title)
+        >>> plt.tight_layout()
+        >>> plt.show()
+        """
+
```

Comment:
this one could be simplified futher by just returning `self.electric_field(xyz_m, xyz_n=xyz_n) / self.rho` the `electric_field` call will take care of the necessary checks for when `xyz_n=None`.